Properly split remote proxy http https

arthurpassos opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->

This PR does 3 things:

1. Fixes small remote resolver bug - more info below.
2. Removes `DB::ProxyConfiguration::Protocol::ANY`
3. Adds proxy docs

Proxy servers must be picked based on the request protocol (i.e, https request should grab the `https_proxy` or whatever `https` related configuration). This was being done correctly for the environment and list proxy resolvers, but not for the remote resolver.

This is how it was implemented:
* The environment resolver would query `http_proxy` or `https_proxy` environment variables based on the request protocol.
* The list resolver would round-robin URIs listed in `<proxy><http OR https></proxy></>` based on the request protocol.
* The remote resolver, on the other hand, would look into the proxy server protocol, not to what the user configured as a resolver for that protocol (i.e, `proxy_scheme`) - **This is wrong**, it's perfectly fine to have a HTTPS request routed through a HTTP proxy. The comparisong that should be made is the same that is being done for list and environment, that is: request protocol x user configuration for that protocol.

This PR addresses this issue by changing the syntax of remote resolvers. It now has to be encapsulated into `<http>` or `<https>` tags just like the list resolver.

```
<http> // new
     <resolver>
           // rest as usual
     </resolver>
</http>
<https> // new
     <resolver>
           // rest as usual
     </resolver>
</https>
```

Not sure I should classify this as bug or improvement.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Select remote proxy resolver based on request protocol, add proxy feature docs and remove `DB::ProxyConfiguration::Protocol::ANY`.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

arthurpassos committed:
tmp

arthurpassos committed:
Merge branch 'master' into properly_split_remote_proxy_http_https

arthurpassos committed:
add some tests

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #55430 from arthurpassos/ClickHouse:properly_split_remote_proxy_http_https -->
*This is an automated comment for commit 63562cbb20ed59032e55548e289c24e3c9a1e0aa with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55430/63562cbb20ed59032e55548e289c24e3c9a1e0aa/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

arthurpassos committed:
trigger ci

arthurpassos committed:
that it?

arthurpassos committed:
fix tests

arthurpassos committed:
remove useless http resolver

arthurpassos commented:
@tavplubix Could you please take a look at this?

arthurpassos commented:
@tavplubix kind ping

Enmk commented on docs/en/operations/server-configuration-parameters/settings.md:
```diff
@@ -2630,3 +2630,114 @@ Possible values:
+            <endpoint>http://resolver:8080/hostname</endpoint>
+            <proxy_scheme>http</proxy_scheme>
+            <proxy_port>80</proxy_port>
+            <proxy_cache_time>10</proxy_cache_time>
```
`time` is too unspecific, please add units like seconds, minutes, hours, etc... into the config node name. Also, what 0 would mean? Cache forever or do not cache at all?

arthurpassos replied:
The old configuration was already like this, changing it will break backwards compatibility. I can modify this for the new version only, but code will be a bit more complex and syntax will be even more different.

I suggest we keep it like this, what do you think?

Enmk commented on docs/en/operations/server-configuration-parameters/settings.md:
```diff
@@ -2630,3 +2630,114 @@ Possible values:
+2. Proxy lists
+3. Environment variables
+
+ClickHouse will check the highest priority resolver type for the request protocol. If it is not defined,
```
What is the condition of CH for checking the next proxy configuration? Would it try proxy lists if the resolver doesn't respond?

arthurpassos replied:
It will check the next proxy configuration if it does not find a configuration for the requests protocol. For instance: ClickHouse is trying to send a HTTPS request and there is a remote resolver configuration section, but only for HTTP. In this case, it will check list and fallback to environment if there is no list for HTTPS.

arthurpassos commented:
@tavplubix kind ping :)

arthurpassos commented:
@tavplubix daily ping :)

tavplubix commented:
The changelog entry is too long and detailed. Please add a short user-readable description for the changelog (one line, a few sentences)

arthurpassos commented:
> The changelog entry is too long and detailed. Please add a short user-readable description for the changelog (one line, a few sentences)

I have moved all text above changelog entry and added a small sentence, could you please check if it is ok now?

tavplubix commented:
Stateless tests (release) - https://github.com/ClickHouse/ClickHouse/pull/55844
Stress test (tsan) - https://github.com/ClickHouse/ClickHouse/pull/55890

tavplubix committed:
trigger CI

arthurpassos commented:
Thanks @tavplubix!
